### PR TITLE
Support ECDSA_P521_SHA512 when using aws_lc_rs feature

### PR DESCRIPTION
**Pull request by Alvenix**
Hope I didn't miss anything.

### CODE REVIEW

**Review by djc — APPROVED** [review]
Looks good to me!

**Comment by cpu** [comment]
I think the `KeyPairAlgorithm` enum in `rustls-cert-gen` could also use an update for the new alg.

**Comment by djc** [comment]
> > I think the `KeyPairAlgorithm` enum in `rustls-cert-gen` could also use an update for the new alg.
> 
> Should I add aws_lc_rs feature to rustls-cert-gen crate?

If necessary, yes!

**Comment by cpu** [comment]
`rustls-cert-gen` was extended with RSA key gen support, which is already tied to using aws-lc-rs. Wouldn't that have required piping through the feature or did I miss something?

**Comment by est31** [comment]
one doesn't *have* to pipe features through, one can also depend on rcgen directly, or do the `cargo install --features rcgen/...` trick. This can download additional dependencies not present in Cargo.toml. scary, right?

**Comment by est31** [comment]
note that in this instance, I think it makes sense to add features to `rustls-cert-gen`.

File: rustls-cert-gen/Cargo.toml [hunk 1]
```diff
@@ -7,13 +7,19 @@ edition.workspace = true
...
+ring = { workspace = true, optional = true }
 rand = { workspace = true }
 anyhow = "1.0.75"
+aws-lc-rs = { version = "1.6.0", optional = true }
 
 [dev-dependencies]
 assert_fs = "1.0.13"
 x509-parser = { workspace = true, features = ["verify"] }
+
+[features]
```
Comment (djc) [inline rustls-cert-gen/Cargo.toml:22]: I think we should move `[features]` above `[dependencies]`.

File: rustls-cert-gen/src/cert.rs [hunk 2]
```diff
@@ -5,6 +5,11 @@ use rcgen::{
 };
 use std::{fmt, fs::File, io, path::Path};
 
+#[cfg(all(feature = "aws_lc_rs", not(feature = "ring")))]
+use aws_lc_rs as ring_like;
+#[cfg(not(all(feature = "aws_lc_rs", not(feature = "ring"))))]
```
Comment (djc) [inline rustls-cert-gen/src/cert.rs:10]: This is overly complicated, just make it `cfg(feature = "ring")`?

File: rustls-cert-gen/src/cert.rs [hunk 3]
```diff
@@ -341,6 +363,25 @@ mod tests {
 		check_signature(&cert, &issuer);
 		Ok(())
 	}
+	#[test]
```
Comment (djc) [inline rustls-cert-gen/src/cert.rs:366]: Nit: leave an empty line above this test, please.

**Review by djc — APPROVED** [review]
Looks okay to me modulo some nits.

File: rustls-cert-gen/Cargo.toml [hunk 4]
```diff
@@ -6,13 +6,19 @@ license.workspace = true
...
 [dependencies]
-rcgen = { path = "../rcgen", default-features = false, features = ["pem", "ring"] }
+rcgen = { path = "../rcgen", default-features = false, features = ["pem"] }
 bpaf = { version = "0.9.5", features = ["derive"] }
 pem = { workspace = true }
-ring = { workspace = true }
+ring = { workspace = true, optional = true }
 rand = { workspace = true }
 anyhow = "1.0.75"
+aws-lc-rs = { version = "1.6.0", optional = true }
```
Comment (cpu) [inline rustls-cert-gen/Cargo.toml:21]: I think we should share the aws-lc-rs version throughout the workspace, like `ring`.

**Review by cpu — APPROVED** [review]
Thanks! Looks good :-)

**Comment by cpu** [comment]
For what it's worth I was able to use the code in this branch over in https://github.com/rustls/rustls/pull/1852 :+1:

File: rcgen/src/sign_algo.rs [hunk 5]
```diff
@@ -177,8 +184,17 @@ pub(crate) mod algo {
 		oid_components: &[1, 2, 840, 10045, 4, 3, 3],
 		params: SignatureAlgorithmParams::None,
 	};
-
-	// TODO PKCS_ECDSA_P521_SHA512 https://github.com/briansmith/ring/issues/824
+	/// ECDSA signing using the P-521 curves and SHA-512 hashing as per [RFC 5758](https://tools.ietf.org/html/rfc5758#section-3.2)
+	/// Currently this is only supported when using aws_lc_rs feature
```
Comment (est31) [inline rcgen/src/sign_algo.rs:188]:
```suggestion
	/// Currently this is only supported with the `aws_lc_rs` feature
```

File: rcgen/src/oid.rs [hunk 6]
```diff
@@ -20,6 +20,10 @@ pub(crate) const EC_PUBLIC_KEY: &[u64] = &[1, 2, 840, 10045, 2, 1];
 pub(crate) const EC_SECP_256_R1: &[u64] = &[1, 2, 840, 10045, 3, 1, 7];
 /// secp384r1 in [RFC 5480](https://datatracker.ietf.org/doc/html/rfc5480#appendix-A)
 pub(crate) const EC_SECP_384_R1: &[u64] = &[1, 3, 132, 0, 34];
+/// secp521r1 in [RFC 5480](https://datatracker.ietf.org/doc/html/rfc5480#appendix-A)
+/// Currently this is only supported when using aws_lc_rs feature
```
Comment (est31) [inline rcgen/src/oid.rs:24]:
```suggestion
/// Currently this is only supported with the `aws_lc_rs` feature
```

**Review by est31 — APPROVED** [review]
LGTM

**Comment by djc** [comment]
IIRC aws-lc-rs with FIPS is only supported on Linux for now.

**Comment by cpu** [comment]
> IIRC aws-lc-rs with FIPS is only supported on Linux for now.

Agreed. There are additional FIPS build reqs [documented here](https://github.com/aws/aws-lc-rs/blob/main/aws-lc-rs/README.md#build) that include "Linux".

> ci / Clippy (-p rcgen --no-default-features) (pull_request) Failing after 4m

I think this is a true-positive finding for this branch:

```
error: unused import: `super::*`
    --> rcgen/src/certificate.rs:1146:6
     |
1146 |     use super::*;
     |         ^^^^^^^^
     |
     = note: `-D unused-imports` implied by `-D warnings`
     = help: to override `-D warnings` add `#[allow(unused_imports)]`
```

**Comment by djc** [comment]
> I got the same warning on the main branch of this repo. That is why I think it is unrelated. Should I fix this warning in this pull request?

Either a separate commit in this PR or a separate PR would be great!

**Comment by cpu** [comment]
> I got the same warning on the main branch of this repo. That is why I think it is unrelated.

Oh! My mistake, sorry about the confusion.

**Comment by cpu** [comment]
Is there anything holding back fixing the CI and rebasing so this could merge? I think we just need the non-Linux CI tasks to avoid the FIPS feature.

**Review by cpu — APPROVED** [review]
Thanks!

File: rcgen/src/key_pair.rs [hunk 7]
```diff
@@ -249,6 +249,18 @@ impl KeyPair {
...
+			#[cfg(all(feature = "aws_lc_rs", not(feature = "ring")))]
+			if alg == &PKCS_ECDSA_P521_SHA512 {
+				KeyPairKind::Ec(ecdsa_from_pkcs8(
+					&signature::ECDSA_P521_SHA512_ASN1_SIGNING,
+					&serialized_der,
+					rng,
+				)?)
+			} else {
+				panic!("Unknown SignatureAlgorithm specified!");
+			}
```
[the remark this answers is not in the document]
Comment (djc) [reply rcgen/src/key_pair.rs:261]: You're right, sorry for the noise!

**Comment by djc** [comment]
Thanks for getting this in good shape!

**Pull request merged**